presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/README.md

```diff
@@ -0,0 +1,148 @@
+
+The plan testing framework attempts to solve the problem of specifiying
+expectations about query plans in tests.
+
+## Quickstart
+
+If you want to see examples of basic use of the plan testing framework, look at
+TestPlanFramework.java. If you'd like to try to begin writing new tests, look
+at PlanMatchPattern.java to see what sorts of patterns are already
+well-supported by the framework.
+
+### PlanMatchPatterns
+
+Expectations are expressed using `PlanMatchPatterns`. A `PlanMatchPattern`
+roughly corresponds to a `PlanNode`, in that it has zero or more
+`PlanMatchPatterns` as sources. A complete test generally expresses its
+expectations about an actual plan as a tree of `PlanMatchPatterns` that is
+structurally similar to the actual plan.
+
+A `PlanMatchPattern` by itself contains no expectations about the contents or
+even type of the node that it's meant to match. Instead, expectations are added
+to `PlanMatchPatterns` using `Matchers`.
+
+### Matchers
+
+`Matchers` are applied to `PlanNodes` to determine if the node they're applied
+to meets some expectation expressed in the `Matcher`.
+
+*A `Matcher` failing to match against a PlanNode it is applied to does not
+intrinsically mean that the plan does not match the expectations expressed by
+the test.* This is because the plan testing framework uses a pattern-matching
+like algorithm to allow tests to express their expectations flexibly with
+regards to changes in the structure of the tree by using the `anyTree()`
+pattern.  The `anyTree()` pattern looks for its child patterns in any of the
+children of the node it's applied to.
+
+Adding matchers to a `PlanMatchPattern` is usually done using one of the static
+methods in PlanMatchPattern.java.
+
+### Aliases
+
+At a high level, Presto plans represent data through `Symbols`. A
+`TableScanNode`, for instance, generates one symbol per column in the table as
+its output.  `ProjectNodes` can apply a calculation to some input `Symbol` and
+generate a new `Symbol` as output.
+
+If you are writing tests, you may care about what is happening to data in the
+actual plan. The simplest possible example of this would be asserting that a
+`SELECT name FROM people` query has an `OutputNode` with an output `Symbol`
+that actually represents the data in the `name` column of the `people` table.
+
+In order for the plan testing framework to be independent of how Presto
+allocates `Symbols`, `Symbols` are given an alias where they originate, and
+must be referenced by that alias further up the plan. The value being assigned
+to a `Symbol` can either be matched against something that is independent of a
+`Symbol` (e.g., a `ColumnHandle` in a `TableScanNode`) or a value that is
+dependent on a previously aliased symbol (e.g., an `Expression` in a
+`ProjectNode`). Matching against the right side of the assignment ensures that
+the plan matching framework is independent of the implementation of the
+`SymbolAllocator`.
+
+## Two types of tests
+
```

Comment:
Suggest deleting this heading and elevating the headings of the two types of tests.